tempesta_fw/http2/huffman.c

```diff
@@ -0,0 +1,641 @@
+						}
+						current += 8;
+						c = Bit_Join8(c, * src++);
+						n--;
+						m--;
+					}
+					else {
+					     /* Last full prefix also processed here */
+					     /* (see current <= 0 above): */
+						buffer_close(source, m);
+						return http2_huffman_decode_tail_fragment(
+							c, destination, current, state, dst, k
+						);
+					}
+				}
+				i = (c >> current) & HT_NMASK;
+				shift = state[i].shift;
+				offset = state[i].offset;
+				if (shift >= 0) {
+					if (Opt_Unlikely(k == 0)) {
+						dst = buffer_expand(destination, &k);
+						if (Opt_Unlikely(k == 0)) {
+							return HTTP2Error_Out_Of_Memory;
+						}
+					}
+					* dst++ = (uchar) offset;
+					k--;
+					current -= shift;
+					goto Root;
+				}
+				else {
+					current += shift;
+				#ifdef HT_BALANCED_TREE
+					if (Opt_Likely(offset)) {
+				#else
+					if (Opt_Likely(offset > 0)) {
+				#endif
+						state = HTDecode + offset;
+						if (offset >= HT_SMALL) {
+							break;
+						}
+					}
+					else {
+						goto End;
+					}
+				}
+			}
+			current += HT_NBITS - HT_MBITS;
+			{
+				fast shift;
+				ufast i;
+				if (current < 0) {
+					if (Opt_Likely(n)) {
+						if (Opt_Unlikely(m == 0)) {
+							src = buffer_next(source, &m);
+						}
+						current += 8;
+						c = Bit_Join8(c, * src++);
+						n--;
+						m--;
+					}
+					else {
+						buffer_close(source, m);
+						return http2_huffman_decode_tail_s_fragment(
+							c, destination, current, state, dst, k
+						);
+					}
+				}
+				i = (c >> current) & HT_MMASK;
+				shift = state[i].shift;
+				offset = state[i].offset;
+				if (Opt_Likely(shift >= 0)) {
+					if (Opt_Unlikely(k == 0)) {
+						dst = buffer_expand(destination, &k);
+						if (Opt_Unlikely(k == 0)) {
+							return HTTP2Error_Out_Of_Memory;
+						}
+					}
+					* dst++ = (uchar) offset;
+					k--;
+					current -= shift;
+				}
+				else {
+					break;
+				}
+			}
+		}
+	     /* Optimization pass in MSVC generates non-optimal code here: */
+		#if defined(HT_BALANCED_TREE) && ! defined(_MSC_VER)
+End:			return HTTP2Error_Huffman_UnexpectedEOS;
+		#else
+End:			return offset == 0 ? HTTP2Error_Huffman_UnexpectedEOS :
+					     HTTP2Error_Huffman_InvalidCode;
+		#endif
+	}
+	else {
+		return 0;
+	}
+}
+
+#ifdef Platform_32bit
+
+#define Write1()		     \
+	dst[0] = (char) (aux >> 24)
+#define Write2()		     \
+	Write1();		     \
+	dst[1] = (char) (aux >> 16)
+#define Write4()		     \
+	Write2();		     \
+	dst[2] = (char) (aux >> 8);  \
+	dst[3] = (char) aux
+#define WriteAux Write4
+
+#else
+
+#define Write1()		     \
+	dst[0] = (char) (aux >> 56)
+#define Write2()		     \
+	Write1();		     \
+	dst[1] = (char) (aux >> 48)
+#define Write4()		     \
+	Write2();		     \
+	dst[2] = (char) (aux >> 40); \
+	dst[3] = (char) (aux >> 32)
+#define WriteAux()		     \
+	Write4();		     \
+	dst[4] = (char) (aux >> 24); \
+	dst[5] = (char) (aux >> 16); \
+	dst[6] = (char) (aux >> 8);  \
+	dst[7] = (char) aux
```

Comment:
And the changes will mean for us a lot of work. However, the scenario is very unlikely and we should not use the complex and slow code just to avoid improbable issues.

In general, there a lot of amazing open source software which is highly portable. However, the software can not utilize full power of a particular OS and a hardware. Tempesta FW targets maximum performance by full utilization of a target platform. Our README.md clearly limits ever x86-64 range of compatible hardware (typically you can not run Tempesta FW on a CPU older than Haswell) and we do a lot of work to get maximum from the hardware.